torch/csrc/distributed/autograd/init.cpp

```diff
@@ -0,0 +1,66 @@
+#include <torch/csrc/distributed/autograd/context/DistAutogradContainer.h>
+#include <torch/csrc/jit/pybind_utils.h>
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/utils/object_ptr.h>
+#include <torch/csrc/utils/pybind.h>
+#include <torch/types.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+namespace {
+
+template <typename T>
+using shared_ptr_class_ = py::class_<T, std::shared_ptr<T>>;
+
+PyObject* dist_autograd_init(PyObject* /* unused */) {
+  auto dist_module =
+      THPObjectPtr(PyImport_ImportModule("torch.distributed.autograd"));
+  if (!dist_module) {
+    throw python_error();
+  }
+
+  auto module = py::handle(dist_module).cast<py::module>();
+
+  auto distAutogradContext =
+      shared_ptr_class_<DistAutogradContext>(module, "DistAutogradContext")
+          .def(
+              "context_id",
+              &DistAutogradContext::context_id,
+              py::call_guard<py::gil_scoped_release>());
+
+  module.def("init", [](int16_t worker_id) {
+    DistAutogradContainer::init(worker_id);
+  });
+
+  module.def("new_context", []() {
+    return DistAutogradContainer::getInstance().newContext();
+  });
+
+  module.def("release_context", [](int64_t context_id) {
+    return DistAutogradContainer::getInstance().releaseContext(context_id);
+  });
+
+  module.def("retrieve_context", [](int64_t context_id) {
+    return DistAutogradContainer::getInstance().retrieveContext(context_id);
+  });
```

Comment:
These need docs. Because they are bound from C++, the current idiom is to bind docstrings to these methods from `torch.distributed.autograd`

Reply:
Hmm, but in this method definition style (where you import a module and attach methods to it) you might have an ordering problem. In any case, we need docs on these methods somehow...

Reply:
If I understand correctly, this Python binding style was copied from the existing c10d code. That's understandable. I'm not a big fan of this binding style though, because it means in a file like `torch/distributed/autograd/__init__.py` you call `release_context` and I'm like, "What the heck, where did this identifier come from?" Essentially, you are monkeypatching these definitions into the module.

A more ordinary and explicable way of exposing modules is to actually define a C binding module (similar to `torch._C`) and then reference this C exposed module explicitly.

Reply:
It also makes writing type stubs easier, since you no longer have to write a type stub for all of `torch/distributed/autograd/__init__.py`, you can just write a type stub for the `_C` module, and then put types inline in `torch/distributed/autograd/__init__.py` (because you're Python 3 only and you can do that, right? :)

Reply:
Thanks for the suggestion, I'll look into what we do for `torch._C` and update the PR accordingly.

Reply:
Looking at torch._C it seems like that binding code (https://github.com/pytorch/pytorch/blob/c5482e33e9ef59af6c18059378964a00036d95aa/torch/csrc/autograd/python_variable.cpp#L538-L621) is much more complicated than what we have here.